src/Type/ObjectProphecy/WillExtendOrImplementDynamicReturnTypeExtension.php

```diff
@@ -92,9 +100,9 @@ public function getTypeFromMethodCall(
             [
                 Type\TypeCombinator::intersect(
                     new Type\ObjectType($className),
-                    ...$calledOnType->getTypes()
+                    //...$calledOnType->templ
```

Comment:
commenting this looks wrong.

I would also suggest trying to remove that extension entirely (without removing the tests) as I think the phpdoc tags in `ObjectProphecy` should be enough to describe that already.

Reply:
Sorry, this is my fault, it's pending in #347 since I didn't know how to do this.

Reply:
@stof 

> I think the phpdoc tags in ObjectProphecy should be enough to describe that already.
I agree with your suggestion. 

sadly but, current prophecy's phpdoc (from https://github.com/phpspec/prophecy/pull/574)  is only about `this-out`.

```php
     * @template U of object
     * @phpstan-param class-string<U> $interface
     * @phpstan-this-out static<T&U>
     */
    public function willImplement($interface)
```

not defined like `@phpstan-return static<T&U>`.

so we got bellow errors when remove this WillExtendOrImplementDynamicReturnTypeExtension removed.

```
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  Line   test/StaticAnalysis/Test/ObjectProphecy/WillImplementTest.php
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  62     Parameter #1 $bar of method JanGregor\Prophecy\Test\StaticAnalysis\Src\BaseModel::bar() expects JanGregor\Prophecy\Test\StaticAnalysis\Src\Bar, JanGregor\Prophecy\Test\StaticAnalysis\Src\Foo given.
         🪪  argument.type
  84     Method JanGregor\Prophecy\Test\StaticAnalysis\Test\ObjectProphecy\WillImplementTest::createProphecy() should return Prophecy\Prophecy\ObjectProphecy<JanGregor\Prophecy\Test\StaticAnalysis\Src\Bar> but returns
         Prophecy\Prophecy\ObjectProphecy<JanGregor\Prophecy\Test\StaticAnalysis\Src\Foo>.
         🪪  return.type
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

- refs. ( debugging with dumpType)

```diff
diff --git a/test/StaticAnalysis/Test/ObjectProphecy/WillImplementTest.php b/test/StaticAnalysis/Test/ObjectProphecy/WillImplementTest.php
index 05b61de..62693c7 100644
--- a/test/StaticAnalysis/Test/ObjectProphecy/WillImplementTest.php
+++ b/test/StaticAnalysis/Test/ObjectProphecy/WillImplementTest.php
@@ -50,7 +50,12 @@ final class WillImplementTest extends Framework\TestCase

     public function testCreateProphecyInTestMethod(): void
     {
+        $prophecyParamOut = $this->prophesize(Src\Foo::class);
+        $prophecyParamOut->willImplement(Src\Bar::class);
+        \PHPStan\dumpType($prophecyParamOut);
+
         $prophecy = $this->prophesize(Src\Foo::class)->willImplement(Src\Bar::class);
+        \PHPStan\dumpType($prophecy);
```


```
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  Line   test/StaticAnalysis/Test/ObjectProphecy/WillImplementTest.php
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  55     Dumped type: Prophecy\Prophecy\ObjectProphecy<JanGregor\Prophecy\Test\StaticAnalysis\Src\Bar&JanGregor\Prophecy\Test\StaticAnalysis\Src\Foo>
  58     Dumped type: Prophecy\Prophecy\ObjectProphecy<JanGregor\Prophecy\Test\StaticAnalysis\Src\Foo>
  67     Parameter #1 $bar of method JanGregor\Prophecy\Test\StaticAnalysis\Src\BaseModel::bar() expects JanGregor\Prophecy\Test\StaticAnalysis\Src\Bar, JanGregor\Prophecy\Test\StaticAnalysis\Src\Foo given.
         🪪  argument.type
  89     Method JanGregor\Prophecy\Test\StaticAnalysis\Test\ObjectProphecy\WillImplementTest::createProphecy() should return Prophecy\Prophecy\ObjectProphecy<JanGregor\Prophecy\Test\StaticAnalysis\Src\Bar> but returns
         Prophecy\Prophecy\ObjectProphecy<JanGregor\Prophecy\Test\StaticAnalysis\Src\Foo>.
         🪪  return.type
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

Reply:
This is actually a bug in phpstan: https://github.com/phpstan/phpstan/issues/8439

Reply:
Nope, that can be solved with a conditional type: https://phpstan.org/r/4bd81d3d-c0cf-40bf-a8a8-9b951e317771

```
 * @return ($class is class-string ? ObjectProphecy<T> : ObjectProphecy<object>)
```

Reply:
Woha, you already came to the same conclusion a year ago :smile: 

https://github.com/phpspec/prophecy/commit/9ab4a3c30c301d2202195f3fef6f0463ed8e18c5#diff-81b4cc8adbc7615b7b55568528ba2dc2436049e7fd1fd1c1567d8d9ed0d234d5R77

Reply:
@stof 
> This is actually a bug in phpstan: https://github.com/phpstan/phpstan/issues/8439
oh, my bad! I had did not understand.

Reply:
at https://github.com/Jan0707/phpstan-prophecy/pull/348/commits/7061ef10554f36409b2ffd8cf7971cec1a57d654

I rolled back the comment-out.
I assume the reason for the code being commented out was that a method-not-found error would be reported by phpstan if the GenericObjectType was not properly asserted. @Jean85